### ISSUE TITLE
missing override check

### ISSUE THREAD

**Issue by olhotak** [lampepfl/dotty#329]
```
abstract class C[T] { def head: T }
case class D[T](head: Int) extends C[T]
```

Notice the type of `head` in `D` is (incorrectly) `Int` instead of `T`.

This compiles in dotty but not in scalac.

**Comment by olhotak** [comment]
The types of the two `head`s are:
`ExprType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),D)),C$$T))`
`TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Int)`

`OverridingPairs` calls `matches` on them and gets `false`. It therefore considers them not to override, so their return types are never checked for consistency.

**Comment by olhotak** [comment]
Here is an even simpler example of the missing check, not involving a case class:
```
abstract class A { def foo: Int }
class B extends A { val foo: String = "foo" }
```

Scalac says:
```
overriding.scala:2: error: overriding method foo in class A of type => Int;
 value foo has incompatible type
class B extends A { val foo: String = "foo" }
                        ^
one error found
```

Dotc does not report any error.

Interestingly, if I change the `val` in `class B` to a `def`, then dotc does report the error.

**Comment by olhotak** [comment]
In the last example, the two non-`match`ing types are:
```
ExprType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Int))
TypeRef(TermRef(ThisType(TypeRef(NoPrefix,scala)),Predef),String)
```

**Comment by olhotak** [comment]
Here is another example where dotty still fails to find an error:
```
class A {
  def m: String = "foo"
}

class B extends A {
  override val m: Int = 42
}

class C extends A {
  override def m: Int = 42
}
```

Scalac finds errors in both class B and class C. Dotty finds an error in class C, but not in class B, where the overriding member is a val instead of a def.

**Comment by odersky** [comment]
Fixed in latest #339. I have added your example as a test.
